src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs

```diff
@@ -0,0 +1,376 @@
+                        indexToEditKinds[edit.NewIndex] = editKindWithoutInsert == default ? SemanticTokenEditKind.Insert : SemanticTokenEditKind.Update;
+                        break;
+                    case EditKind.Delete:
+                        indexToEditKinds.TryGetValue(edit.OldIndex, out var editKindWithoutDelete);
+                        Contract.ThrowIfTrue(editKindWithoutDelete == SemanticTokenEditKind.Delete); // We shouldn't have two deletions at the same position
+                        indexToEditKinds[edit.OldIndex] = editKindWithoutDelete == default ? SemanticTokenEditKind.Delete : SemanticTokenEditKind.Update;
+                        break;
+                }
+            }
+
+            return CombineEditsIfPossible(newGroupedSemanticTokens, indexToEditKinds);
+        }
+
+        private static SemanticTokensEdit[] CombineEditsIfPossible(
+            SemanticToken[] newGroupedSemanticTokens,
+            Dictionary<int, SemanticTokenEditKind> indexToEditKinds)
+        {
+            // This method combines the edits into the minimal possible edits (for the most part).
+            // For example, if an index contains both an insertion and deletion, we combine the two
+            // edits into one.
+            // We also combine edits if we have consecutive edits of the same types, i.e.
+            // Delete->Delete, Insert->Insert, and Update->Update.
+            // Technically, we could combine Update->Insert, and Update->Delete, but those cases have
+            // special rules and would complicate the logic. They also generally do not result in a
+            // huge reduction in the total number of edits, so we leave them out for now.
+
+            using var _ = ArrayBuilder<LSP.SemanticTokensEdit>.GetInstance(out var semanticTokensEdits);
+
+            var editIndices = indexToEditKinds.Keys.ToArray();
+
+            // The indices in indexToEdit kinds are not guaranteed to be in chronological order when we
+            // extract them from the dictionary. We must sort the edit kinds by index since we need to
+            // know what kind of edits surround a given index in order to potentially combine them into
+            // one edit.
+            Array.Sort(editIndices);
+
+            // Example to give clarity to orderedEditNumber and orderedTokenNumber variables defined below:
+            // Non-grouped semantic tokens: 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14
+            // orderedEditNumber:           0                   1
+            // orderedTokenNumber:          0         1         2
+            for (var orderedEditNumber = 0; orderedEditNumber < editIndices.Length; orderedEditNumber++)
+            {
+                var orderedTokenNumber = editIndices[orderedEditNumber];
+                var initialEditKind = indexToEditKinds[orderedTokenNumber];
```

Comment:
```suggestion
                var orderedTokenNumber = editIndices[orderedEditNumber];
                var initialEditKind = indexToEditKinds[orderedTokenNumber];
                var editStartPoistion = orderedTokenNumber * 5;
```

(and then just use this below in all cases).